dbms/src/Storages/Transaction/Collator.h

```diff
@@ -91,4 +119,119 @@ extern TiDBCollators dummy_collators;
 extern std::vector<std::string> dummy_sort_key_contaners;
 extern std::string dummy_sort_key_contaner;
 
+struct TiDBCollatorPtrImpl : TiDBCollatorID
+{
+    class ITiDBCollator const * ptr{};
+
+    TiDBCollatorPtrImpl(class ITiDBCollator const * ptr_, int32_t collator_id_)
+        : TiDBCollatorID(collator_id_)
+        , ptr(ptr_)
+    {
+    }
+
+    explicit TiDBCollatorPtrImpl(class ITiDBCollator const * ptr_)
+        : TiDBCollatorID(ptr_ ? ptr_->getCollatorId() : 0)
+        , ptr(ptr_)
+    {
+    }
+
+    ALWAYS_INLINE inline int compare(const char * s1, size_t length1, const char * s2, size_t length2) const
+    {
+        if (likely(canUseFastPath()))
+        {
+            return DB::BinCollatorCompare<true>(s1, length1, s2, length2);
+        }
+        else
+        {
+            return compareIndirect(s1, length1, s2, length2);
+        }
+    }
+
+    ALWAYS_INLINE inline int compareIndirect(const char * s1, size_t length1, const char * s2, size_t length2) const
+    {
+        return ptr->compare(s1, length1, s2, length2);
```

Comment:
Don't need to check if `ptr` is nullptr?

Reply:
No need to check. Just like previous way, the caller must guarantee the pointer `class ITiDBCollator const *` is not null. `TiDBCollatorPtrImpl` is a wrap for raw poiner.

Reply:
rollback those modifications.